hyperledger/sawtooth-core
# Enforce and abide by PoET block claim frequency (zTest) restrictions

Per the PoET specification, validators are limited in how frequently they may claim blocks - i.e., the zTest.  If it is determined that allowing a validator to win an election (i.e., claim a block and have it added to the blockchain) would cause it to claim blocks more frequently than a configurable statistical rate, the block is rejected.

This PR updates:
* adds a fixed duration block count property to the PoET configuration view.  This value indicates the number of blocks which will have their local mean computed by using a simple ratio of target to initial wait times.  These blocks are not considered when performing the zTest.
* updates the PoET enclave factory to use the new PoET configuration view's fixed duration block count property.
* adds zTest block counts to both the consensus and validator state objects.  The zTest block count is the number of blocks claimed by all validators and a specific validator, respectively, that are covered by the zTest.  Additionally, code is added to properly maintain these values.
* adds minimum win count and maximum win deviation properties to the PoET configuration view.  These values are used when performing the zTest.  The minimum win count is that minimum number of blocks, after the fixed duration block count has been surpassed, a validator has to win before zTest will apply to any further blocks.  The maximum win deviation is used to determine if a validator has violated the zTest policy by comparing this against the deviation from the expected number of blocks a validator should have won.
* updates the PoET block verifier to ensure that when verifying blocks it checks to make sure a validator is throttled by the block claim frequency allowed 
* updates the PoET block publisher to refuse to initialize a block if the block frequency check fails, indicating that other validators would reject the block
* refactors all PoET policy tests (key block claim limit, block claim delay, and block claim frequency)
* adds a short-circuit for blocks upon which block publisher has already determined it would not be allowed to build on top of

To test locally:
 `cd /project/sawtooth-core/integration/sawtooth_integration/docker` 

In `poet-smoke.yaml`, after the line:
`sawtooth.consensus.algorithm=poet \`

Add:

```
sawtooth.poet.target_wait_time=5 \
sawtooth.poet.initial_wait_time=0 \
sawtooth.poet.block_claim_delay=1 \
sawtooth.poet.fixed_duration_block_count=5 \
sawtooth.poet.ztest_maximum_win_deviation=1.645 \
sawtooth.poet.ztest_minimum_win_count=2 \
```

This should increase the chance of validators refusing the publish blocks because it would violate the block claim frequency test.  You can look at the logs for error log messages.

To reduce the likelihood of zTest failures, you can change `sawtooth.poet.ztest_maximum_win_deviation`.  The values are based upon the confidence interval (i.e., how confident we are that we have truly detected a validator winning at a frequency we consider too frequent):

```
3.075 => 99.9%
2.575 => 99.5%
2.321 => 99%
1.645 => 95%
```

Execute:
`run_docker_test poet-smoke.yaml`

Go get your beverage of choice and cross your fingers that the test passes.
